### ISSUE TITLE
[v3 ETH Mainnet] Fill or Kill Order failing due to insufficient gas 

### ISSUE THREAD

**Issue by wizardlabsxyz** [trader-xyz/nft-swap-sdk#105]
Why would this contract function run out of gas?

Example of a transaction that failed from my application
https://etherscan.io/tx/0x628c7c14d69e838a05ac7dbd8e5b4e6f968eae810c7940ca42ddeb4e02947428

**Comment by johnrjj** [comment]
I'm guessing it has to do with one of the NFTs requiring a bit more gas than usual to do a transfer (sometimes there are custom transfer hooks and methods)

I confirmed it is indeed a gas issue -- https://dashboard.tenderly.co/tx/mainnet/0x628c7c14d69e838a05ac7dbd8e5b4e6f968eae810c7940ca42ddeb4e02947428

I re-simulated your transaction with a 250k gas limit and it went through successfully -- 
<img width="1168" alt="Screen Shot 2023-06-28 at 9 54 49 PM" src="https://github.com/trader-xyz/nft-swap-sdk/assets/1103963/02628374-9403-4cb5-889f-1401d67e5b11">

From there I profiled the gas and it looks like the first NFT transferred takes extra gas in the transfer method of the contract:
<img width="1154" alt="Screen Shot 2023-06-28 at 9 55 03 PM" src="https://github.com/trader-xyz/nft-swap-sdk/assets/1103963/113ef90d-688b-4908-a7bd-fafa454941a2">

I would just pad this transaction with a bit more gas.

**Comment by jinsley8** [comment]
I'm having the same issue, can we set a higher gas limit like this?

https://github.com/trader-xyz/nft-swap-sdk/blob/5782d110b8882f7490f33b80acf59cd6f2ea7301/test/v3/swap.test.ts#L81

**Comment by wizardlabsxyz** [comment]
I solved this by adding a gas buffer and this essentially doubles the gas limit           
```
const fillTx = await nftSwapSdk.fillSignedOrder(signedOrder, { gasAmountBufferMultiple: 2 });
```

**Comment by jinsley8** [comment]
Default `fillSignedOrder` seems to run out of gas for certain NFT collections for single swaps ERC721<>ERC721 plus fees:
```js
const fillTx = await swapSdk.fillSignedOrder(normalizedOrder);
```

This gets over the limit for the same transaction:
```js
const fillTx = await swapSdk.fillSignedOrder(normalizedOrder, undefined, {
    gasLimit: '250000',
});
```
However, if I try swapping 3 ERC721 for 1 ERC721 the gas is like 420,000.

I'll test `gasAmountBufferMultiple`, otherwise I may try calculating gas based on number of assets in the swap. 250,000 base for 2, plus 80,000-100,000 for each additional asset.

Until I ran into this issue, I had been thinking if the txn hash is returned then it was successful. I think you be better to wait for a couple of block confirmations to determine if the txn is a success or failed.

I'm thinking of using wagmi `waitForTransaction` to wait for a couple block confirmations before continuing with the rest of my code.

https://wagmi.sh/react/hooks/useWaitForTransaction
```js
import { useWaitForTransaction } from 'wagmi'

const fillTx = await swapSdk.fillSignedOrder(normalizedOrder);
const fillTxReceipt = await swapSdk.awaitTransactionHash(fillTx.hash);
console.log(`useHandleApproveTrade: 🎉 🥳 Order filled. TxHash: ${fillTxReceipt.transactionHash}`);

const waitForTransaction = useWaitForTransaction({
    chainId: 1,
    confirmations: 2,
    hash: fillTxReceipt.transactionHash,
    onSuccess(data) {
      console.log('Success', data)
    },
    onError(error) {
      console.log('Error', error)
    },
})
```

**Comment by wizardlabsxyz** [comment]
does the result of `awaitTransactionHash` not indicate whether the transaction failed or not?

I've used `useWaitForTransaction` before and my assumption was that `awaitTransactionHash` was the same thing. It sounds like you're saying that `awaitTransactionHash` actually resolves before the transaction is actually completed in some cases. Is that what you mean?

**Comment by jinsley8** [comment]
It returns `TransactionReceipt` as:
```js
const fillTxReceipt = {
    accessList: null
    blockHash: null
    blockNumber: null
    chainId: 0
    confirmations: 0
    creates: null
    data: "0xe14b5....00000000000000000000000000000000000000000000000000"
    from: "0x0A...A0"
    to: "0x0...94"
    hash: "0x9eb...29"
    gasLimit: BigNumber {_hex: '0x0493e0', _isBigNumber: true}
    gasPrice: BigNumber {_hex: '0x2439710915', _isBigNumber: true}
    maxFeePerGas: BigNumber {_hex: '0x2439710915', _isBigNumber: true}
    maxPriorityFeePerGas: BigNumber {_hex: '0x094d4705eb', _isBigNumber: true}
    nonce: 248
    r: "0x213...a66"
    s: "0x7f7...b9bba3"
    transactionIndex: null
    type: 2
    v: 0
    value: BigNumber {_hex: '0x00', _isBigNumber: true}
    wait: (confirmations) => {…}
}
```

`useWaitForTransaction` returns that as data plus state:
https://wagmi.sh/react/hooks/useWaitForTransaction#return-value

**Comment by wizardlabsxyz** [comment]
it looks like fillTxReceipt has a function called wait that takes an integer as the number of confirmations you want to wait for. Perhaps using fillTxReceipt.wait(2) is the same as using useWaitForTransaction as you illustrated above.

**Comment by jinsley8** [comment]
Yes, possibly, I'll test it out. It looks like it returns errors.

https://docs.ethers.org/v5/api/providers/types/#providers-TransactionResponse

**Comment by jinsley8** [comment]
Okay this works:
```js
const fillTx = await swapSdk.fillSignedOrder(normalizedOrder, { gasAmountBufferMultiple: 1.4 });
const waitForConfirmation = await fillTx.wait(2);
```

It returns [TransactionReceipt](https://docs.ethers.org/v5/api/providers/types/#providers-TransactionReceipt) if successul otherwise it errors.
